
ING.JOSEARREGOCES




IMPORTAR LIBRERIAS

In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,precision_score,recall_score,confusion_matrix
from sklearn.metrics import roc_curve, auc
from sklearn import metrics 
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn import metrics
from sklearn.metrics import matthews_corrcoef
from sklearn.preprocessing import Normalizer

DATOS

In [6]:
datosTitanic = pd.read_csv('/content/train.csv')
datosTitanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

ELIMINACION COLUMNAS INNECESARIAS

In [7]:
datosTitanic = datosTitanic.drop(columns = ['Cabin','PassengerId','Name','Ticket'],axis=1)
datosTitanic

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Embarked
0           0       3    male  22.0      1      0   7.2500        S
1           1       1  female  38.0      1      0  71.2833        C
2           1       3  female  26.0      0      0   7.9250        S
3           1       1  female  35.0      1      0  53.1000        S
4           0       3    male  35.0      0      0   8.0500        S
..        ...     ...     ...   ...    ...    ...      ...      ...
886         0       2    male  27.0      0      0  13.0000        S
887         1       1  female  19.0      0      0  30.0000        S
888         0       3  female   NaN      1      2  23.4500        S
889         1       1    male  26.0      0      0  30.0000        C
890         0       3    male  32.0      0      0   7.7500        Q

[891 rows x 8 columns]

REEMPLAZO DE VALORES SIMPLIFICADOS

In [8]:
datosTitanic.replace({'Embarked':{'S':0,'C':1,'Q':2}},inplace=True)
datosTitanic.replace({'Sex':{'male':0,'female':1}},inplace=True)
datosTitanic

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0           0       3    0  22.0      1      0   7.2500       0.0
1           1       1    1  38.0      1      0  71.2833       1.0
2           1       3    1  26.0      0      0   7.9250       0.0
3           1       1    1  35.0      1      0  53.1000       0.0
4           0       3    0  35.0      0      0   8.0500       0.0
..        ...     ...  ...   ...    ...    ...      ...       ...
886         0       2    0  27.0      0      0  13.0000       0.0
887         1       1    1  19.0      0      0  30.0000       0.0
888         0       3    1   NaN      1      2  23.4500       0.0
889         1       1    0  26.0      0      0  30.0000       1.0
890         0       3    0  32.0      0      0   7.7500       2.0

[891 rows x 8 columns]

AMPLIACION DE VALORES

In [9]:
datosTitanic['Embarked'].fillna(datosTitanic['Embarked'].mode()[0], inplace=True)
datosTitanic['Pclass'].fillna(datosTitanic['Pclass'].mode()[0], inplace=True)
datosTitanic['Sex'].fillna(datosTitanic['Sex'].mode()[0], inplace=True)
datosTitanic['Age'].fillna(datosTitanic['Age'].mode()[0], inplace=True)
datosTitanic['SibSp'].fillna(datosTitanic['SibSp'].mode()[0], inplace=True)
datosTitanic['Parch'].fillna(datosTitanic['Parch'].mean(), inplace=True)
datosTitanic['Fare'].fillna(datosTitanic['Fare'].mean(), inplace=True)
datosTitanic

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0           0       3    0  22.0      1      0   7.2500       0.0
1           1       1    1  38.0      1      0  71.2833       1.0
2           1       3    1  26.0      0      0   7.9250       0.0
3           1       1    1  35.0      1      0  53.1000       0.0
4           0       3    0  35.0      0      0   8.0500       0.0
..        ...     ...  ...   ...    ...    ...      ...       ...
886         0       2    0  27.0      0      0  13.0000       0.0
887         1       1    1  19.0      0      0  30.0000       0.0
888         0       3    1  24.0      1      2  23.4500       0.0
889         1       1    0  26.0      0      0  30.0000       1.0
890         0       3    0  32.0      0      0   7.7500       2.0

[891 rows x 8 columns]

NORMALIZACION

In [10]:
norma = Normalizer().fit(datosTitanic)
normalizado=norma.transform(datosTitanic)
normalizado

array([[0.        , 0.12832206, 0.        , ..., 0.        , 0.31011166,
        0.        ],
       [0.01237466, 0.01237466, 0.01237466, ..., 0.        , 0.88210632,
        0.01237466],
       [0.03651957, 0.10955871, 0.03651957, ..., 0.        , 0.28941759,
        0.        ],
       ...,
       [0.        , 0.08881718, 0.02960573, ..., 0.05921145, 0.6942543 ,
        0.        ],
       [0.0251657 , 0.0251657 , 0.        , ..., 0.        , 0.75497086,
        0.0251657 ],
       [0.        , 0.09057442, 0.        , ..., 0.        , 0.23398392,
        0.06038295]])

DATA FRAME

In [11]:
dataframe=pd.DataFrame(normalizado ,columns=['Sobrevivio','ClasePasajero','Sexo','Edad','Hermanos','Padres','Costo','Embarco'])
dataframe

Sobrevivio  ClasePasajero      Sexo      Edad  Hermanos    Padres  \
0      0.000000       0.128322  0.000000  0.941028  0.042774  0.000000   
1      0.012375       0.012375  0.012375  0.470237  0.012375  0.000000   
2      0.036520       0.109559  0.036520  0.949509  0.000000  0.000000   
3      0.015716       0.015716  0.015716  0.550066  0.015716  0.000000   
4      0.000000       0.083243  0.000000  0.971173  0.000000  0.000000   
..          ...            ...       ...       ...       ...       ...   
886    0.000000       0.066593  0.000000  0.899002  0.000000  0.000000   
887    0.028127       0.028127  0.028127  0.534417  0.000000  0.000000   
888    0.000000       0.088817  0.029606  0.710537  0.029606  0.059211   
889    0.025166       0.025166  0.000000  0.654308  0.000000  0.000000   
890    0.000000       0.090574  0.000000  0.966127  0.000000  0.000000   

        Costo   Embarco  
0    0.310112  0.000000  
1    0.882106  0.012375  
2    0.289418  0.000000  
3    0.834529  0.000000  
4    0.223370  0.000000  
..        ...       ...  
886  0.432853  0.000000  
887  0.843816  0.000000  
888  0.694254  0.000000  
889  0.754971  0.025166  
890  0.233984  0.060383  

[891 rows x 8 columns]

In [12]:
Y=datosTitanic['Survived']
X=dataframe.drop(columns=['Sobrevivio'], axis=1)
X

ClasePasajero      Sexo      Edad  Hermanos    Padres     Costo   Embarco
0         0.128322  0.000000  0.941028  0.042774  0.000000  0.310112  0.000000
1         0.012375  0.012375  0.470237  0.012375  0.000000  0.882106  0.012375
2         0.109559  0.036520  0.949509  0.000000  0.000000  0.289418  0.000000
3         0.015716  0.015716  0.550066  0.015716  0.000000  0.834529  0.000000
4         0.083243  0.000000  0.971173  0.000000  0.000000  0.223370  0.000000
..             ...       ...       ...       ...       ...       ...       ...
886       0.066593  0.000000  0.899002  0.000000  0.000000  0.432853  0.000000
887       0.028127  0.028127  0.534417  0.000000  0.000000  0.843816  0.000000
888       0.088817  0.029606  0.710537  0.029606  0.059211  0.694254  0.000000
889       0.025166  0.000000  0.654308  0.000000  0.000000  0.754971  0.025166
890       0.090574  0.000000  0.966127  0.000000  0.000000  0.233984  0.060383

[891 rows x 7 columns]

In [13]:
Y

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

In [14]:
Xtrain, Xtest, Ytrain, Ytest = train_test_split(X, Y, random_state=4)

KNN

In [26]:
ClasificadorKNN = KNeighborsClassifier()
ClasificadorKNN.fit(Xtrain,Ytrain)
Ypred = ClasificadorKNN.predict(Xtest)
PresicionKNN = round(accuracy_score(Ytest, Ypred), 6)
print("Presicion:",PresicionKNN*100,"%")
print("MCC:",matthews_corrcoef(Ytest, Ypred))

Presicion: 74.8879 %
MCC: 0.4297716894977169


REGRESION

In [16]:
regresion = LogisticRegression()
regresion.fit(Xtrain,Ytrain)
Ypred1 = regresion.predict(Xtrain)
Presicion = round(metrics.accuracy_score(Ytrain, Ypred1), 5)
PorcentajePresicion = 100 * Presicion 
print("PResicion:", PorcentajePresicion,"%")
print("MCC:",matthews_corrcoef(Ytrain, Ypred1))

PResicion: 68.413 %
MCC: 0.322975083738602


In [17]:
regresion.fit(Xtest,Ytest)
Ypred2 = regresion.predict(Xtest)
Presicion1 = round(metrics.accuracy_score(Ytest, Ypred2), 6)
PorcentajePresicion1 = 100 * Presicion1 
print("Presicion:", PorcentajePresicion1,"%")
print("MCC:",matthews_corrcoef(Ytest, Ypred2))

Presicion: 71.3004 %
MCC: 0.2888290005259896


SVM

In [18]:
fichero = SVC(kernel='linear')
fichero.fit(Xtrain, Ytrain)
Ypred3 = fichero.predict(Xtest)
Presicion2 = round(metrics.accuracy_score(Ytest, Ypred3),6)
PorcentajePresicion2 = 100 * Presicion2 
print("presicion:", PorcentajePresicion2,"%")
print("MCC:",matthews_corrcoef(Ytest, Ypred3))

presicion: 71.7489 %
MCC: 0.3309665393008779


SE ESCOGE KNN TRAS SER EL METODO MAS EFECTIVO CON PRESICION DEL 74.8879 %

In [19]:
Prueba = pd.read_csv('/content/test.csv')
Prueba.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [20]:
Prueba = Prueba.drop(columns = ['Cabin','PassengerId','Name','Ticket'],axis=1)
Prueba['Embarked'].fillna(Prueba['Embarked'].mode()[0], inplace=True)
Prueba['Pclass'].fillna(Prueba['Pclass'].mode()[0], inplace=True)
Prueba['Sex'].fillna(Prueba['Sex'].mode()[0], inplace=True)
Prueba['Age'].fillna(Prueba['Age'].mode()[0], inplace=True)
Prueba['SibSp'].fillna(Prueba['SibSp'].mode()[0], inplace=True)
Prueba['Parch'].fillna(Prueba['Parch'].mean(), inplace=True)
Prueba['Fare'].fillna(Prueba['Fare'].mean(), inplace=True)
Prueba.replace({'Embarked':{'S':0,'C':1,'Q':2}},inplace=True)
Prueba.replace({'Sex':{'male':0,'female':1}},inplace=True)
Prueba

Pclass  Sex   Age  SibSp  Parch      Fare  Embarked
0         3    0  34.5      0      0    7.8292         2
1         3    1  47.0      1      0    7.0000         0
2         2    0  62.0      0      0    9.6875         2
3         3    0  27.0      0      0    8.6625         0
4         3    1  22.0      1      1   12.2875         0
..      ...  ...   ...    ...    ...       ...       ...
413       3    0  21.0      0      0    8.0500         0
414       1    1  39.0      0      0  108.9000         1
415       3    0  38.5      0      0    7.2500         0
416       3    0  21.0      0      0    8.0500         0
417       3    0  21.0      1      1   22.3583         1

[418 rows x 7 columns]

In [21]:
normaTest = Normalizer().fit(Prueba)
normalizadoTest=normaTest.transform(Prueba)
dataFrameTest=pd.DataFrame(normalizadoTest ,columns=['ClasePasajero','Sexo','Edad','Hermanos','Padres','Costo','Embarco'])
dataFrameTest

ClasePasajero      Sexo      Edad  Hermanos    Padres     Costo   Embarco
0         0.084363  0.000000  0.970179  0.000000  0.000000  0.220166  0.056242
1         0.062980  0.020993  0.986690  0.020993  0.000000  0.146954  0.000000
2         0.031839  0.000000  0.987010  0.000000  0.000000  0.154220  0.031839
3         0.105212  0.000000  0.946909  0.000000  0.000000  0.303800  0.000000
4         0.117944  0.039315  0.864921  0.039315  0.039315  0.483078  0.000000
..             ...       ...       ...       ...       ...       ...       ...
413       0.132221  0.000000  0.925548  0.000000  0.000000  0.354793  0.000000
414       0.008644  0.008644  0.337120  0.000000  0.000000  0.941343  0.008644
415       0.076353  0.000000  0.979859  0.000000  0.000000  0.184519  0.000000
416       0.132221  0.000000  0.925548  0.000000  0.000000  0.354793  0.000000
417       0.097185  0.000000  0.680295  0.032395  0.032395  0.724297  0.032395

[418 rows x 7 columns]

APLICACION DE KNN A PRUEBA

In [22]:
YpredTest = ClasificadorKNN.predict(dataFrameTest)
YpredTest

array([0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1,
       1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0,
       1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0,

In [25]:
final=pd.DataFrame(YpredTest ,columns=['Survived'])
idtit = pd.read_csv('/content/gender_submission.csv')
Id=idtit['PassengerId']
result=pd.concat([Id, final], axis=1)
result.columns=['PassengerId', 'Survived']
result.to_csv('Prediction.csv', index=False)